analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs

```diff
@@ -63,21 +75,103 @@ public override void ExecutionCompleted()
+        foreach (var (operation, block) in alwaysFalse.Select(x => (x.Key, x.Value)))
         {
-            ReportIssue(operation, false);
+            ReportIssue(operation, block, false);
         }
 
         base.ExecutionCompleted();
     }
 
-    private void ReportIssue(IOperation operation, bool conditionEvaluation)
+    private void ReportIssue(IOperation operation, BasicBlock block, bool conditionIsTrue)
+    {
+        var issueMessage = operation.Kind == OperationKindEx.IsNull ? MessageNullCoalescing : string.Format(MessageBool, conditionIsTrue);
+        var secondaryLocations = SecondaryLocations(block, conditionIsTrue);
+        if (secondaryLocations.Any())
+        {
+            ReportIssue(Rule2583, operation, secondaryLocations, string.Format(MessageUnreachable, issueMessage));
+        }
+        else
+        {
+            ReportIssue(Rule2589, operation, null, string.Format(issueMessage, conditionIsTrue));
+        }
+    }
+
+    private List<Location> SecondaryLocations(BasicBlock block, bool conditionIsTrue)
+    {
+        List<Location> locations = new();
+        IOperation currentStart = null;
+        IOperation currentEnd = null;
+        var unreachable = UnreachableOperations(block, conditionIsTrue).Where(x => !IsIgnoredLocation(x.Syntax));
+        foreach (var operation in unreachable.Concat(reached).OrderBy(x => x.Syntax.SpanStart))
+        {
+            if (unreachable.Contains(operation))
+            {
+                currentStart ??= operation;
+                currentEnd = operation;
+            }
+            else
+            {
+                if (currentStart is not null)
+                {
+                    locations.Add(currentStart.Syntax.CreateLocation(currentEnd.Syntax));
+                    currentStart = null;
+                }
+            }
+        }
+        if (currentStart != null)
+        {
+            locations.Add(currentStart.Syntax.CreateLocation(currentEnd.Syntax));
+        }
+        return locations;
+    }
+
+    private IEnumerable<IOperation> UnreachableOperations(BasicBlock block, bool conditionIsTrue)
     {
-        var issueMessage = operation.Kind == OperationKindEx.IsNull ? MessageNullCoalescing : string.Format(MessageBool, conditionEvaluation);
-        ReportIssue(Rule2589, operation, null, issueMessage);
+        if (block.SuccessorBlocks.Distinct().Count() != 2)
+        {
+            return Enumerable.Empty<IOperation>();
+        }
+        HashSet<BasicBlock> reachable = new() { block };
+        HashSet<BasicBlock> unreachable = new();
+
+        var reachableSuccessor = conditionIsTrue ^ block.ConditionKind == ControlFlowConditionKind.WhenFalse
+            ? block.ConditionalSuccessor.Destination
+            : block.SuccessorBlocks.Single(x => x != block.ConditionalSuccessor.Destination);
+        TraverseReachable(reachableSuccessor);
+        var unreachableSuccessor = block.SuccessorBlocks.Single(x => x != reachableSuccessor);
```

Comment:
I'm not happy with the way this variable is calculated - as a side-effect of the previous computation on a collection. Because we could directly take the "other" value. What about this 
```
        var conditionalIsReachable = (block.ConditionKind == ControlFlowConditionKind.WhenTrue) == conditionIsTrue;
        TraverseReachable(conditionalIsReachable ? block.ConditionalSuccessor.Destination : block.FallThroughSuccessor.Destination);
        TraverseUnreachable(conditionalIsReachable ? block.FallThroughSuccessor.Destination : block.ConditionalSuccessor.Destination);
```
And we could get rid of the 4x `.Destination` if we change the parameter to branch. And we'd need to add nullcheck for `Destination`, because that one can be null

Reply:
`Destination` cannot be null for a successor of a block that triggers `ConditionEvaluated`.
I agree that it makes sense to check for null to satisfy the API and be safe for future changes, but it will be impossible to write a test covering it.

Reply:
As it is now, the `if (block.SuccessorBlocks.Distinct().Count() != 2)` makes sure that `.Destination` is not null so we do not need the null check